### DISCUSSION TITLE
What changed between Symfony 3 -> 4 that could break auto-wiring of factory services?

### DISCUSSION THREAD

**Discussion by Bilge** [symfony/symfony#43969] [Q&A] [answered]
Currently upgrading Symfony 3.4 to 4.4 and I'm getting compiler pass errors that would seem to indicate that services that are clearly defined, and were previously working, are not being picked up. 

>  [Symfony\Component\DependencyInjection\Exception\RuntimeException]
  Cannot autowire service "Org\MyBundle\Repository\MyRepository": argument "$class" of method "Doctrine\ORM\EntityRepository::__construct()" references class "Doctrine\ORM\Mapping\ClassMetadata" but no such service exists.

This error indicates it is trying to inject the service on its own without using the factory definition we've created.

~~My best guess is because the service being injected is declared by a different bundle than the one requesting it. This is given that this project is still structured using the very old approach of putting all first-party code in various "bundles" despite it being all one application. Given that the entire application _is_ in dozens of bundles, it would be impractical to change this structure, so in case my inclination is correct, is there any way to move forward with this upgrade without restructuring the project?~~

---

Update: it seems my previous guess was incorrect, because if I copy the service definition directly into the bundle consuming it, there is no change. The only way I can get it working again is if I manually specify the service for the parameter type hinted as `Org\MyBundle\Repository\MyRepository`; it no longer matches the class name from the type hint of the constructor parameter to the service definition with the same ID. Here is my service definition that no longer can be autowired, for reasons that remain unclear to me:

```xml
        <service id="Org\MyBundle\Repository\MyRepository">
            <factory service="doctrine.orm.entity_manager" method="getRepository" />
            <argument>Org\MyBundle\Entity\MyEntity</argument>
        </service>
```

I am vaguely aware that sometimes the compiler complains when you specify a factory but not the class it creates, so I tried adding the `class` attribute to both the `<service>` and `<factory>` tags, but it still will not inject this service via auto-wiring any more.

---

### Brainstorm

* Creating a service definition for the Controller and specifying the repository's class name as the service ID for the repository parameter **works**, even though the service ID matches the type hint already and auto-wiring is turned on, and this configuration worked on Symfony 3.
* Could it be to do with the order in which bundle definition files are parsed? (Probably not, the debug session shows all services are present and correct at the time the exception is thrown).
* Could it be to do with "bundle inheritance", since that feature was removed? Don't know what it is or if we used that.
* Could it be to do with the Kernel flags that changed between v3 and v4, since Flex already updated the Kernel for us. (Probably not, reverting doesn't make any difference).

### Debug session

Placing a breakpoint on the exception at [`DefinitionErrorExceptionPass:54`](https://github.com/symfony/symfony/blob/8e59e85f0a9310036d22ae62db326172b86779c4/src/Symfony/Component/DependencyInjection/Compiler/DefinitionErrorExceptionPass.php#L54), I browsed `$graph->nodes` to see if I could find the repository service definition, and I did.

![image](https://user-images.githubusercontent.com/470626/140892910-4ec92306-5d61-49c8-b28b-6c41d461e046.png)

This service looks exactly as we would expect, with the factory set up the way we want and having its in-edges source nodes all pointing to the services that throw these errors. And yet, the value of `$value` in this context is just a blank service definition for `MyRepository`, with the factory missing.

![image](https://user-images.githubusercontent.com/470626/140893833-0374911e-4c24-495f-99a9-b5a01ee83bf5.png)

@nicolas-grekas May I ask for your input on this?

**Comment by ahundiak** [comment, marked as answer]
Still not sure I understand the issue.

Repositories extending the ServiceEntityRepository can be autowired because the entity class name is actually specified in the constructor.  Repositories extending EntityRepository need to have the entity class name defined in the service definition.  Pretty much always been that way.

**Reply by Bilge** [reply]
My repositories extend `EntityRepository` and are declared as I have shown in the XML configuration. In v3 they can be injected using repository type hints and in v4 they cannot. I do not know why this is.

**Reply by ahundiak** [reply]
And do you have a _defaults with autowire true section in each of your service files?  

You mentioned in a previous question that you have lots of bundles.  Did you by any chance add an application level service file with a ./src resource section?  It could be a case where autowire is trying to create the same service twice.

And App\MyRepository is a somewhat strange namespace.  I assume you edited what you posted?  Are you sure the actual namespaces match?

**Reply by Bilge** [reply]
Good questions. Each bundle's configuration section starts with the following.

```xml
        <defaults autowire="true" autoconfigure="true" public="false"/>

        <prototype namespace="App\MyBundle\" resource="../../*" />
```

There is not any application level service file imposing its own resource/prototype section. Again, the configuration was valid and working in v3, and such a configuration where resources were being overridden in the manner you describe would of course be incorrect and invalid. The configuration only broke down on upgrading to v4.4.

Yes, all the namespaces have been edited for confidentiality reasons.

**Reply by stof** [reply]
Repositories not extending ServiceEntityRepository should not be defined as services in the first place, as that would produce weird bugs when resetting the entity manager (as your service would be the repository associated with the old entity manager). The ServiceEntityRepository class is precisely here to solve that case (by instantiating the repository with a reference to the EntityManager proxy instance, not the internal instance, and so stable across resetting).

Regarding your service definition, this `<prototype>` will define its own service for the repository (without using the factory) as it matches the file. So make sure that your own explicit service definition comes after, so that it wins.

**Reply by Bilge** [reply]
>make sure that your own explicit service definition comes after, so that it wins.

It does. Of course it does, otherwise it wouldn't have worked on v3. The `<defaults>` and `<prototype>` tags are always the first two tags under `<services>` in each bundle. So unless XML ordering broke in v4 then that cannot be the issue.

Regarding `EntityRepository` and `ServiceEntityRepository`, I have no idea what that is about; I guess I can look into it in due course, however it does not appear to be the issue at present.

**Reply by stof** [reply]
Well, in 3.4, were you defining your service with id `Org\MyBundle\Repository\MyRepository` and the factory or not at all ?

**Reply by Bilge** [reply]
Config has not changed one bit. The service definition we were and are using is the one I described above: (you can also see this in the active Symfony 4.4 Xdebug session)

```xml
        <service id="Org\MyBundle\Repository\MyRepository">
            <factory service="doctrine.orm.entity_manager" method="getRepository" />
            <argument>Org\MyBundle\Entity\MyEntity</argument>
        </service>
```

I understand this may not be strictly correct for the "resetting" feature (whatever that is), but I guess we are not using that feature since it would be broken anyway.

**Reply by Bilge** [reply]
I just want to highlight this part from the "brainstorm" section:

>Creating a service definition for the Controller and specifying the repository's class name as the service ID for the constructor's repository parameter **works**, even though the service ID matches the type hint already and auto-wiring is turned on, and this configuration worked on Symfony 3.

That is, if I manually force my repository definition into the controller, it's fine. It just isn't using the existing service definition when doing auto-wiring. And we know the existing service definition is present and correct due to the Xdebug session, it just declines to use it for some reason, even though the type hint does match the service ID exactly.

**Reply by stof** [reply]
Well, your debugging sessions shows an invalid definition of the `Org\MyBundle\Repository\MyRepository` service, as it defines that service without a factory, and so will try to autowire the constructor arguments of the `MyRepository` class and will fail.

Can you provide code snippets showing your working and your non-working cases ? Understanding the difference will be easier.

**Reply by Bilge** [reply]
Yes, the _second_ screenshot shows the factory is missing when it tries to inject the service, hence the error, _but at the same time_, I confirmed the correct service definition _does exist_ in the node graph; it just declines to use it. Why does it do that?

I'm not sure what you mean by working and non-working cases. I cannot get the container to compile, so as far as that goes, I do not know what is working; only what isn't working (at the point it throws an exception)? If you mean, the case where I can fix the issue, I fixed it this way:

```xml
    <service id="Org\MyBundle\Controller\MyController">
        <argument type="service" key="$myRepository" id="Org\MyBundle\Repository\MyRepository" />
    </service>
```

Essentially, I force it to use the service by matching the parameter name (manually) instead of the parameter type (automatically, with auto-wiring), since parameter type injection appears to be broken, but only for these repositories, since there are another 4 non-repository services that I guess are injected just fine since it moves on to another controller after fixing this one in this way.

**Reply by Bilge** [reply]
Bleargh, it's so gross. It was because of a case mismatch. The class name had a capital `I` and the service id had a lower-case `i`. Apparently Symfony 3 ignores case and Symfony 4 does not. I am sorry for wasting your time, since, given I was censoring the service names (due to confidentiality agreements) you would never be able to spot the error. Nevertheless, thanks @stof and @ahundiak for your help, it helped me to realize the problem _had to be_ something to do with the type hint, and indeed it was a case mismatch.

**Reply by stof** [reply]
Indeed, service ids are now case sensitive, as defined in https://github.com/symfony/symfony/blob/4.0/UPGRADE-4.0.md#dependencyinjection

> - Service identifiers and parameter names are now case sensitive.

**Reply by Bilge** [reply]
It is easy to miss that single bullet in a long list of bullets within five pages of upgrade nodes (4.0, 4.1, 4.2, 4.3, 4.4), but as it happens, I had actually seen it, but even then it's difficult to know that the problem you're looking at is related to that change.

**Reply by faizanakram99** [reply]
As a thumb rule, upgrade regularly. These issues occur only when upgrading from LTS to LTS.
